File: Source/GlueGenerator/PropertyTranslators/SinglecastDelegatePropertyTranslator.cpp
#include "SinglecastDelegatePropertyTranslator.h"

#include "GlueGenerator/CSGenerator.h"
#include "GlueGenerator/CSScriptBuilder.h"

bool FSinglecastDelegatePropertyTranslator::CanHandleProperty(const FProperty* Property) const
{
	const FDelegateProperty* DelegateProperty = CastField<FDelegateProperty>(Property);

	if (!FCSGenerator::Get().CanExportFunctionParameters(DelegateProperty->SignatureFunction))
	{
		return false;	
	}
	
	return true;
}

void FSinglecastDelegatePropertyTranslator::AddDelegateReferences(const FProperty* Property, TSet<UFunction*>& DelegateSignatures) const
{
	const FDelegateProperty* DelegateProperty = CastField<FDelegateProperty>(Property);
	DelegateSignatures.Add(DelegateProperty->SignatureFunction);
}

FString FSinglecastDelegatePropertyTranslator::GetManagedType(const FProperty* Property) const
{
	return GetDelegateName(CastFieldChecked<FDelegateProperty>(Property));
}

void FSinglecastDelegatePropertyTranslator::ExportPropertyStaticConstruction(FCSScriptBuilder& Builder, const FProperty* Property, const FString& NativePropertyName) const
{
	FDelegateBasePropertyTranslator::ExportPropertyStaticConstruction(Builder, Property, NativePropertyName);

	const FDelegateProperty* DelegateProperty = CastFieldChecked<FDelegateProperty>(Property);

	if (DelegateProperty->SignatureFunction->NumParms > 0)
	{
		FString DelegateName = GetDelegateName(DelegateProperty);
		Builder.AppendLine(FString::Printf(TEXT("%s.InitializeUnrealDelegate(%s_NativeProperty);"), *DelegateName, *NativePropertyName));
	}
}

void FSinglecastDelegatePropertyTranslator::ExportMarshalToNativeBuffer(FCSScriptBuilder& Builder, const FProperty* Property, const FString& Owner, const FString& PropertyName, const FString& DestinationBuffer, const FString& Offset, const FString& Source) const
{
	FString DelegateName = GetDelegateName(CastFieldChecked<FDelegateProperty>(Property));
	Builder.AppendLine(FString::Printf(TEXT("DelegateMarshaller<%s>.ToNative(IntPtr.Add(%s, %s), 0, %s, %s);"), *DelegateName, *DestinationBuffer, *Offset, *Owner, *Source));
}

void FSinglecastDelegatePropertyTranslator::ExportCleanupMarshallingBuffer(FCSScriptBuilder& Builder, const FProperty* ParamProperty, const FString& NativeParamName) const
{
	
}

void FSinglecastDelegatePropertyTranslator::ExportMarshalFromNativeBuffer(FCSScriptBuilder& Builder, const FProperty* Property, const FString& Owner, const FString& PropertyName, const FString& AssignmentOrReturn, const FString& SourceBuffer, const FString& Offset, bool bCleanupSourceBuffer, bool reuseRefMarshallers) const
{
	FString DelegateName = GetDelegateName(CastFieldChecked<FDelegateProperty>(Property));
	Builder.AppendLine(FString::Printf(TEXT("%s DelegateMarshaller<%s>.FromNative(IntPtr.Add(%s, %s), IntPtr.Zero, 0, %s);"), *AssignmentOrReturn, *DelegateName, *SourceBuffer, *Offset, *Owner));
}

FString FSinglecastDelegatePropertyTranslator::GetNullReturnCSharpValue(const FProperty* ReturnProperty) const
{
	return "null";
}

FString FSinglecastDelegatePropertyTranslator::GetDelegateName(const FDelegateProperty* Property)
{
	const UFunction* Function = Property->SignatureFunction;
	return FDelegateBasePropertyTranslator::GetDelegateName(Function);
}

File: Source/UnrealSharpScriptGenerator/PropertyTranslators/SinglecastDelegatePropertyTranslator.cs
using EpicGames.UHT.Types;
using UnrealSharpScriptGenerator.Utilities;

namespace UnrealSharpScriptGenerator.PropertyTranslators;

public class SinglecastDelegatePropertyTranslator : DelegateBasePropertyTranslator
{
    public SinglecastDelegatePropertyTranslator() : base(EPropertyUsageFlags.Parameter)
    {
    }
    
    public override bool CanExport(UhtProperty property)
    {
        UhtDelegateProperty delegateProperty = (UhtDelegateProperty) property;
        bool hasReturnValue = delegateProperty.Function.ReturnProperty != null;
        return ScriptGeneratorUtilities.CanExportParameters(delegateProperty.Function) && !hasReturnValue;
    }
    
    public override string GetManagedType(UhtProperty property)
    {
        return $"TDelegate<{GetFullDelegateName(((UhtDelegateProperty) property).Function)}>";
    }

    public override void ExportPropertyStaticConstructor(GeneratorStringBuilder builder, UhtProperty property, string nativePropertyName)
    {
        base.ExportPropertyStaticConstructor(builder, property, nativePropertyName);
        
        UhtDelegateProperty delegateProperty = (UhtDelegateProperty) property;
        
        if (!delegateProperty.Function.HasParameters)
        {
            return;
        }
        
        string fullDelegateName = GetFullDelegateName(delegateProperty.Function, true);
        builder.AppendLine($"{fullDelegateName}.InitializeUnrealDelegate({nativePropertyName}_NativeProperty);");
    }

    public override void ExportToNative(GeneratorStringBuilder builder, UhtProperty property, string propertyName, string destinationBuffer,
        string offset, string source)
    {
        UhtDelegateProperty delegateProperty = (UhtDelegateProperty) property;
        string fullDelegateName = GetFullDelegateName(delegateProperty.Function);
        builder.AppendLine($"SingleDelegateMarshaller<{fullDelegateName}>.ToNative(IntPtr.Add({destinationBuffer}, {offset}), 0, {source});");
    }

    public override void ExportFromNative(GeneratorStringBuilder builder, UhtProperty property, string propertyName,
        string assignmentOrReturn, string sourceBuffer, string offset, bool bCleanupSourceBuffer, bool reuseRefMarshallers)
    {
        UhtDelegateProperty delegateProperty = (UhtDelegateProperty) property;
        string fullDelegateName = GetFullDelegateName(delegateProperty.Function);
        builder.AppendLine($"{assignmentOrReturn} SingleDelegateMarshaller<{fullDelegateName}>.FromNative(IntPtr.Add({sourceBuffer}, {offset}), 0);");
    }

    public override string GetNullValue(UhtProperty property)
    {
        return "null";
    }

    public override void ExportCleanupMarshallingBuffer(GeneratorStringBuilder builder, UhtProperty property,
        string paramName)
    {
        
    }
}